# 약정보, 약 영양소 정보, 권고사항 정보 병합

In [1]:
import pandas as pd

In [3]:
med = pd.read_csv(r'C:\Users\USER\Desktop\med_list.csv')

In [4]:
ingd = pd.read_csv(r'C:\Users\USER\Downloads\ingd_list (1).csv')

In [2]:
info = pd.read_csv(r'C:\Users\USER\Desktop\info_list.csv')

In [17]:
filtered_rows12 = pd.read_csv(r'C:\Users\USER\filtered_rows12.csv', encoding='cp949')

In [18]:
filtered_rows13 = filtered_rows12.dropna()

# med 약 정보 

In [6]:
med

Unnamed: 0          dcode                    dname
0               0  A11AKP08C0036           10%포도당가엔에이케이주1
1               1  A11AKP08C0038           10%포도당가엔에이케이주2
2               2  A11APPPPP0629           10%포도당가엔에이케이주3
3               3  A11AOOOOO2364            5%포도당가엔에이케이주1
4               4  A11AOOOOO2365  5%포도당가엔에이케이주2(1000mL/백)
...           ...            ...                      ...
40601       41658  2020021400003     한풍인삼패독산연조엑스(단미엑스혼합제)
40602       41664  A11A0090A0183                한풍황련탕엑스과립
40603       41672  A11AKP08G3442            해옹엑스과립(연교패독산)
40604       41691  A11AKP08G3535                  황련탕엑스과립
40605       41693  2023100500002                   후로팡설하정

[40606 rows x 3 columns]

In [7]:
med[['extracted']] = med['dname'].str.extract(r'([가-힣]+)')

# 똑같은 약 중복 없애기 filtered_med

In [9]:
unique_dnames = med['dname'].str.extract(r'([가-힣]+)').drop_duplicates()


filtered_med = med[med['dname'].str.extract(r'([가-힣]+)').isin(unique_dnames)[0]]


display(filtered_med)

Unnamed: 0          dcode                 dname    extracted
0               0  A11AKP08C0036        10%포도당가엔에이케이주1   포도당가엔에이케이주
12             12  2018061800005                 가그린목액        가그린목액
13             13  A11AOOOOO8955                  가나릴정         가나릴정
14             14  A11AKP08G3016                 가나메드정        가나메드정
15             15  2020042400051          가나모티에스알정15mg     가나모티에스알정
...           ...            ...                   ...          ...
40601       41658  2020021400003  한풍인삼패독산연조엑스(단미엑스혼합제)  한풍인삼패독산연조엑스
40602       41664  A11A0090A0183             한풍황련탕엑스과립    한풍황련탕엑스과립
40603       41672  A11AKP08G3442         해옹엑스과립(연교패독산)       해옹엑스과립
40604       41691  A11AKP08G3535               황련탕엑스과립      황련탕엑스과립
40605       41693  2023100500002                후로팡설하정       후로팡설하정

[30109 rows x 4 columns]

# ingd 약의 영양소 

In [11]:
ingd

Unnamed: 0 ingd_code ingd_name          dcode
0               0   I001722     무수포도당  A11AKP08C0036
1               1   I003769     염화나트륨  A11AKP08C0036
2               2   I003201      염화칼륨  A11AKP08C0036
3               3   I001722     무수포도당  A11AKP08C0038
4               4   I003769     염화나트륨  A11AKP08C0038
...           ...       ...       ...            ...
98443       98443   I006323        대추  A11AKP08G3535
98444       98444   I006469        반하  A11AKP08G3535
98445       98445   I001699        인삼  A11AKP08G3535
98446       98446   I006174        황련  A11AKP08G3535
98447       98447   I002999   플로로글루시놀  2023100500002

[98448 rows x 4 columns]

# info 약의 영양소를 섭취할 때 권고사항

In [12]:
info

Unnamed: 0                              sunb_name  \
0               0               무수포도당(Glucose Anhydrous)   
1               1                 염화나트륨(Sodium Chloride)   
2               2               염화칼륨(Potassium Chloride)   
3               3     벤지다민염산염(Benzydamine Hydrochloride)   
4               4    세틸피리디늄염화물(Cetylpyridinium Chloride)   
...           ...                                    ...   
15409       15409    티클로피딘염산염(Ticlopidine Hydrochloride)   
15410       15410    티클로피딘염산염(Ticlopidine Hydrochloride)   
15411       15411  히드로코르티손발레레이트(Hydrocortisone Valerate)   
15412       15412       토포테칸염산염(Topotecan Hydrochloride)   
15413       15413       베탁솔롤염산염(Betaxolol Hydrochloride)   

                                         contents_origin ingd_code  
0                                                    NaN   I001722  
1                                                    NaN   I003769  
2                                                    NaN   I003201  
3                                                    NaN   I005942  
4                                                    NaN   I004558  
...                                                  ...       ...  
15409  녹차는 혈전 생성을 억제하고 동맥경화를 예방하는 작용을 합니다. 따라서 항응고제 복...   I005477  
15410  코엔자임 Q10은 혈액응고효과가 있어, 복용시 이 약의 효과가 약간 떨어질 수 있어...   I005477  
15411                                                NaN   I004803  
15412                                                NaN   I005559  
15413  음주를 삼가세요. 음주시 이 약의 부작용(심장박동과 호흡의 느려짐 등)이 심해질 수...   I004540  

[15414 rows x 4 columns]

# 권고사항이 없는 영양소 없애기 

In [13]:
nonna_info = info.dropna()

In [22]:
nonna_info

Unnamed: 0                                    sunb_name  \
5               5             이토프리드염산염(Itopride Hydrochloride)   
6               6  모사프리드시트르산염이수화물(Mosapride Citrate Dihydrate)   
9               9                            오메프라졸(Omeprazole)   
10             10                            오메프라졸(Omeprazole)   
11             11                            오메프라졸(Omeprazole)   
...           ...                                          ...   
15407       15407          티클로피딘염산염(Ticlopidine Hydrochloride)   
15408       15408          티클로피딘염산염(Ticlopidine Hydrochloride)   
15409       15409          티클로피딘염산염(Ticlopidine Hydrochloride)   
15410       15410          티클로피딘염산염(Ticlopidine Hydrochloride)   
15413       15413             베탁솔롤염산염(Betaxolol Hydrochloride)   

                                         contents_origin ingd_code  
5                              식전 30분~1시간 또는 식후에 복용하세요.    I005050  
6                                   식사와 관계없이 복용할 수 있어요.    I002400  
9                           식전 30분~1시간에 복용하는 것이 가장 좋아요.    I002684  
10     음주와 흡연을 삼가세요. 알코올과 니코틴은 위벽을 자극하여 위궤양, 가슴쓰림, 위산...   I002684  
11     {{카페인}} 복용시 주의하세요. 카페인은 하부식도괄약근을 느슨하게 하여 위산역류 ...   I002684  
...                                                  ...       ...  
15407  마늘, 생강, 인삼, 은행엽은 피를 묽게 하여 출혈의 위험성을 높일 수 있으니 과량...   I005477  
15408  과도한 비타민 A와 비타민 E의 병용시 피를 너무 묽게 할 수 있어요. 비타민 E는...   I005477  
15409  녹차는 혈전 생성을 억제하고 동맥경화를 예방하는 작용을 합니다. 따라서 항응고제 복...   I005477  
15410  코엔자임 Q10은 혈액응고효과가 있어, 복용시 이 약의 효과가 약간 떨어질 수 있어...   I005477  
15413  음주를 삼가세요. 음주시 이 약의 부작용(심장박동과 호흡의 느려짐 등)이 심해질 수...   I004540  

[1140 rows x 4 columns]

# nonna_info를 쓸데없는거 날리고 전처리한게 filtered_rows13

In [20]:
filtered_rows13

Unnamed: 0                                  sunb_name  \
0            11                          오메프라졸(Omeprazole)   
1            12                          오메프라졸(Omeprazole)   
2            13                          오메프라졸(Omeprazole)   
3            14                          오메프라졸(Omeprazole)   
4            19                탄산수소나트륨(Sodium Bicarbonate)   
..          ...                                        ...   
303       15316  메틸페니데이트염산염(Methylphenidate Hydrochloride)   
304       15317  메틸페니데이트염산염(Methylphenidate Hydrochloride)   
305       15376           파조파닙염산염(Pazopanib Hydrochloride)   
306       15407        티클로피딘염산염(Ticlopidine Hydrochloride)   
307       15408        티클로피딘염산염(Ticlopidine Hydrochloride)   

                 contents_origin ingd_code  
0                     카페인 복용시 주의   I002684  
1    마늘, 양파 등 자극성이 강한 식품은 섭취에 주의   I002684  
2                      비타민 B의 고갈   I002684  
3                      칼슘 흡수를 저해   I002684  
4                     카페인 복용시 주의   I003757  
..                           ...       ...  
303                   카페인 복용시 주의   I002247  
304              마황 함유 식품 복용에 주의   I002247  
305               자몽주스 섭취를 삼가세요.   I012349  
306        마늘, 생강, 인삼, 은행엽 섭취 주의   I005477  
307          비타민 A와 비타민 E는 적게 섭취   I005477  

[302 rows x 4 columns]

# filtered_rows13의 같은 성분을 한 행에 , 으로 join한게 combined_text

In [38]:
combined_text = filtered_rows13.groupby(['sunb_name', 'ingd_code'])['contents_origin'].apply(lambda x: ' '.join(x)).reset_index()

In [57]:
combined_text

sunb_name ingd_code  \
0                      L-카르니틴(L-Carnitine)   I006336   
1           건조황산제일철(Dried Ferrous Sulfate)   I001118   
2           결합형에스트로겐(Conjugated Estrogens)   I000643   
3    글루콘산제일철수화물(Ferrous Gluconate Hydrate)   I001473   
4                      글리메피리드(Glimepiride)   I001710   
..                                     ...       ...   
169         플루복사민말레산염(Fluvoxamine Maleate)   I001586   
170     플루옥세틴염산염(Fluoxetine Hydrochloride)   I001556   
171  피오글리타존염산염(Pioglitazone Hydrochloride)   I003047   
172        황산철수화물(Ferrous Sulfate Hydrate)   I001478   
173        히드로클로로티아지드(Hydrochlorothiazide)   I004791   

                                       contents_origin  
0                        고단백식이 시, 이 약의 흡수가 저해될 수 있어요.   
1              비타민 C와 복용하면 철분 흡수 향상 육류/어류/가금류는 흡수를 높여줌  
2                                          자몽주스 섭취에 주의  
3              비타민 C와 복용하면 철분 흡수 향상 육류/어류/가금류는 흡수를 높여줌  
4    고탄수화물 식품 섭취 주의 녹차, 아스파라거스 추출물, 민들레 등과 같은 천연 이뇨...  
..                                                 ...  
169                                카페인 복용시 주의 멜라토닌 피하기  
170                                         카페인 복용시 주의  
171  고탄수화물 식품 섭취 주의 녹차, 아스파라거스 추출물, 민들레 등과 같은 천연 이뇨...  
172            비타민 C와 복용하면 철분 흡수 향상 육류/어류/가금류는 흡수를 높여줌  
173                칼륨이 풍부한 식단은 도움을 줌 칼슘 농도를 증가 시킬 수 있음  

[174 rows x 3 columns]

# 권고사항이 있는 영양소들 ---> 11797

권고사항이 있는 영양소들 추출한게 matching_info_rows

In [28]:
matching_info_rows = ingd[ingd['ingd_code'].isin(combined_text['ingd_code'])]

In [31]:
matching_info_rows

Unnamed: 0 ingd_code    ingd_name          dcode
50             50   I002684        오메프라졸  2017080100052
51             51   I003757      탄산수소나트륨  2017080100052
52             52   I002684        오메프라졸  2015062200004
53             53   I003757      탄산수소나트륨  2015062200004
82             82   I003385    슈도에페드린염산염  A11A3100A0067
...           ...       ...          ...            ...
98049       98049   I001414         펠로디핀  A11AOOOOO8873
98060       98060   I004695         시메티딘  A11A0340A0323
98061       98061   I004695         시메티딘  A11A0340A0352
98068       98068   I000566  클로피도그렐황산수소염  A11AOOOOO9132
98091       98091   I001414         펠로디핀  A11AOOOOO6516

[11797 rows x 4 columns]

# 부작용이 있는 약들의 모임 matching_med_rows

권고사항이 있는 영양소들을 가진 약들의 목록 matching_med_rows

In [29]:
matching_med_rows = filtered_med[filtered_med['dcode'].isin(matching_info_rows['dcode'])]

In [30]:
matching_med_rows

Unnamed: 0          dcode            dname extracted
24             24  2017080100052  가나플럭스정20/1100mg    가나플럭스정
45             45  A11A3100A0067             가네카정      가네카정
79             79  2019100100003       가두에정5/10mg      가두에정
82             82  2016011100002  가드메트정100/1000mg     가드메트정
87             87  A11A5070A0028              가딘정       가딘정
...           ...            ...              ...       ...
40507       41301  A11AOOOOO4235           하이드로핀정    하이드로핀정
40511       41313  A11AOOOOO8873            하이디핀정     하이디핀정
40522       41350  A11A0340A0323            하이메틴정     하이메틴정
40528       41374  A11AOOOOO9132            하이빅스정     하이빅스정
40551       41473  A11AOOOOO6516        하이퍼지속정5mg    하이퍼지속정

[7898 rows x 4 columns]

# matching_med_rows 약에  해당하는 ingd_code 컬럼 추가하기

하나의 약이지만 여러가지 성분을 가진 약들이 있기 때문에 행이 늘어났다.

In [32]:
merged_df = pd.merge(matching_med_rows, ingd[['dcode', 'ingd_code']], on='dcode', how='left')

In [33]:
merged_df

Unnamed: 0          dcode            dname extracted ingd_code
0              24  2017080100052  가나플럭스정20/1100mg    가나플럭스정   I002684
1              24  2017080100052  가나플럭스정20/1100mg    가나플럭스정   I003757
2              45  A11A3100A0067             가네카정      가네카정   I003385
3              45  A11A3100A0067             가네카정      가네카정   I005661
4              79  2019100100003       가두에정5/10mg      가두에정   I012151
...           ...            ...              ...       ...       ...
21441       41301  A11AOOOOO4235           하이드로핀정    하이드로핀정   I001414
21442       41313  A11AOOOOO8873            하이디핀정     하이디핀정   I001414
21443       41350  A11A0340A0323            하이메틴정     하이메틴정   I004695
21444       41374  A11AOOOOO9132            하이빅스정     하이빅스정   I000566
21445       41473  A11AOOOOO6516        하이퍼지속정5mg    하이퍼지속정   I001414

[21446 rows x 5 columns]

# 늘어난 행들이 있으니까 dcode가 같으면 ingd_code를 한 행에 넣기

7898--> 21446 ---> 7898

In [34]:
merged_df['ingd_code'] = merged_df.groupby('dcode')['ingd_code'].transform(lambda x: ', '.join(x.dropna()))

merged_df2 = merged_df.drop_duplicates(subset='dcode')

In [35]:
merged_df2

Unnamed: 0          dcode            dname extracted         ingd_code
0              24  2017080100052  가나플럭스정20/1100mg    가나플럭스정  I002684, I003757
2              45  A11A3100A0067             가네카정      가네카정  I003385, I005661
4              79  2019100100003       가두에정5/10mg      가두에정  I012151, I004212
6              82  2016011100002  가드메트정100/1000mg     가드메트정  I002178, I012730
8              87  A11A5070A0028              가딘정       가딘정           I003478
...           ...            ...              ...       ...               ...
21441       41301  A11AOOOOO4235           하이드로핀정    하이드로핀정           I001414
21442       41313  A11AOOOOO8873            하이디핀정     하이디핀정           I001414
21443       41350  A11A0340A0323            하이메틴정     하이메틴정           I004695
21444       41374  A11AOOOOO9132            하이빅스정     하이빅스정           I000566
21445       41473  A11AOOOOO6516        하이퍼지속정5mg    하이퍼지속정           I001414

[7898 rows x 5 columns]

In [41]:
combined_text

sunb_name ingd_code  \
0                      L-카르니틴(L-Carnitine)   I006336   
1           건조황산제일철(Dried Ferrous Sulfate)   I001118   
2           결합형에스트로겐(Conjugated Estrogens)   I000643   
3    글루콘산제일철수화물(Ferrous Gluconate Hydrate)   I001473   
4                      글리메피리드(Glimepiride)   I001710   
..                                     ...       ...   
169         플루복사민말레산염(Fluvoxamine Maleate)   I001586   
170     플루옥세틴염산염(Fluoxetine Hydrochloride)   I001556   
171  피오글리타존염산염(Pioglitazone Hydrochloride)   I003047   
172        황산철수화물(Ferrous Sulfate Hydrate)   I001478   
173        히드로클로로티아지드(Hydrochlorothiazide)   I004791   

                                       contents_origin  
0                        고단백식이 시, 이 약의 흡수가 저해될 수 있어요.   
1              비타민 C와 복용하면 철분 흡수 향상 육류/어류/가금류는 흡수를 높여줌  
2                                          자몽주스 섭취에 주의  
3              비타민 C와 복용하면 철분 흡수 향상 육류/어류/가금류는 흡수를 높여줌  
4    고탄수화물 식품 섭취 주의 녹차, 아스파라거스 추출물, 민들레 등과 같은 천연 이뇨...  
..                                                 ...  
169                                카페인 복용시 주의 멜라토닌 피하기  
170                                         카페인 복용시 주의  
171  고탄수화물 식품 섭취 주의 녹차, 아스파라거스 추출물, 민들레 등과 같은 천연 이뇨...  
172            비타민 C와 복용하면 철분 흡수 향상 육류/어류/가금류는 흡수를 높여줌  
173                칼륨이 풍부한 식단은 도움을 줌 칼슘 농도를 증가 시킬 수 있음  

[174 rows x 3 columns]

In [55]:
merged_df2['merged_contents'] = merged_df2['ingd_code'].apply(
    lambda codes: ', '.join(combined_text.loc[combined_text['ingd_code'].astype(str).isin(map(str, codes.split(','))), 'contents_origin'])
)

C:\Users\USER\AppData\Local\Temp\ipykernel_6212\294403580.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  merged_df2['merged_contents'] = merged_df2['ingd_code'].apply(


In [56]:
merged_df2

Unnamed: 0          dcode            dname extracted         ingd_code  \
0              24  2017080100052  가나플럭스정20/1100mg    가나플럭스정  I002684, I003757   
2              45  A11A3100A0067             가네카정      가네카정  I003385, I005661   
4              79  2019100100003       가두에정5/10mg      가두에정  I012151, I004212   
6              82  2016011100002  가드메트정100/1000mg     가드메트정  I002178, I012730   
8              87  A11A5070A0028              가딘정       가딘정           I003478   
...           ...            ...              ...       ...               ...   
21441       41301  A11AOOOOO4235           하이드로핀정    하이드로핀정           I001414   
21442       41313  A11AOOOOO8873            하이디핀정     하이디핀정           I001414   
21443       41350  A11A0340A0323            하이메틴정     하이메틴정           I004695   
21444       41374  A11AOOOOO9132            하이빅스정     하이빅스정           I000566   
21445       41473  A11AOOOOO6516        하이퍼지속정5mg    하이퍼지속정           I001414   

                                         merged_contents  
0      카페인 복용시 주의 마늘, 양파 등 자극성이 강한 식품은 섭취에 주의 비타민 B의 ...  
2                                             카페인 복용시 주의  
4                                          자몽주스 섭취 피하세요.  
6      고탄수화물 식품 섭취 주의 녹차, 아스파라거스 추출물, 민들레 등과 같은 천연 이뇨...  
8       카페인 복용시 주의 마늘, 양파 등 자극성이 강한 식품은 섭취에 주의 비타민 B의 고갈  
...                                                  ...  
21441                       고지방, 고탄수화물 주의 자몽주스 섭취를 피하세요.  
21442                       고지방, 고탄수화물 주의 자몽주스 섭취를 피하세요.  
21443   카페인 복용시 주의 마늘, 양파 등 자극성이 강한 식품은 섭취에 주의 비타민 B의 고갈  
21444          마늘, 생강, 인삼, 은행엽 섭취 주의 비타민 A와 비타민 E는 적게 섭취  
21445                       고지방, 고탄수화물 주의 자몽주스 섭취를 피하세요.  

[7898 rows x 6 columns]

In [58]:
merged_df2.to_csv('C:/Users/User/merged_df2.csv', index=False)

In [60]:
merged_df2 = merged_df2.drop('extracted', axis=1)

# 병합한 데이터



In [61]:
merged_df2

Unnamed: 0          dcode            dname         ingd_code  \
0              24  2017080100052  가나플럭스정20/1100mg  I002684, I003757   
2              45  A11A3100A0067             가네카정  I003385, I005661   
4              79  2019100100003       가두에정5/10mg  I012151, I004212   
6              82  2016011100002  가드메트정100/1000mg  I002178, I012730   
8              87  A11A5070A0028              가딘정           I003478   
...           ...            ...              ...               ...   
21441       41301  A11AOOOOO4235           하이드로핀정           I001414   
21442       41313  A11AOOOOO8873            하이디핀정           I001414   
21443       41350  A11A0340A0323            하이메틴정           I004695   
21444       41374  A11AOOOOO9132            하이빅스정           I000566   
21445       41473  A11AOOOOO6516        하이퍼지속정5mg           I001414   

                                         merged_contents  
0      카페인 복용시 주의 마늘, 양파 등 자극성이 강한 식품은 섭취에 주의 비타민 B의 ...  
2                                             카페인 복용시 주의  
4                                          자몽주스 섭취 피하세요.  
6      고탄수화물 식품 섭취 주의 녹차, 아스파라거스 추출물, 민들레 등과 같은 천연 이뇨...  
8       카페인 복용시 주의 마늘, 양파 등 자극성이 강한 식품은 섭취에 주의 비타민 B의 고갈  
...                                                  ...  
21441                       고지방, 고탄수화물 주의 자몽주스 섭취를 피하세요.  
21442                       고지방, 고탄수화물 주의 자몽주스 섭취를 피하세요.  
21443   카페인 복용시 주의 마늘, 양파 등 자극성이 강한 식품은 섭취에 주의 비타민 B의 고갈  
21444          마늘, 생강, 인삼, 은행엽 섭취 주의 비타민 A와 비타민 E는 적게 섭취  
21445                       고지방, 고탄수화물 주의 자몽주스 섭취를 피하세요.  

[7898 rows x 5 columns]